github_workflow.md

```diff
@@ -0,0 +1,153 @@
+# GitHub Milestone and Issue Workflow
+
+This purpose of this document is to describe our development workflow on
+GitHub; in particular, how we use milestones and issues. We recognize that our
+many subprojects may use variations of these ideas. This particular workflow is
+based on the practices from the JupyterLab repo
+(https://github.com/jupyterlab/jupyterlab) and are optional for Jupyter
+subprojects to follow.
+
+## Milestones
+
+We use Github milestones to track issues and pull requests related to particular
+releases of a subproject (https://github.com/jupyterlab/jupyterlab/milestones).
+
+* These milestones should be labeled with the version numbers of the releases
+  (`4.3`, `5.0`, etc.).
+* Subprojects should only have a few numbered milestones. Beyond one or two
+  future releases, we have found that detailed milestones become meaningless.
+* Subprojects should also have a `Backlog` milestone for issues and pull
+  requests that have not been assigned to a particular release milestone. This
+  also includes new feature requests ("wishlist") issues that we haven't
+  committed to implementing. Essentially, `Backlog` means "at some point in the
+  future, maybe."
```

Comment:
Backlog here is used more in the sense of an agile approach where *all* new feature requests are put in the "backlog" and then during each planning round, the backlog is reviewed and things are moved to a particular milestone. It is sort of a rename of wishlist, but with a caveta. In the past, we tended to just dump all issue that weren't going to make it into the next release into future *numbered* milestones. That is why the notebook 5.0 milestone has ~150 issues. It isn't that we are actually going to do all 150 of them, we have just been putting all "future work" on that milestone. The difficulty with that approach is that we have found it to be very difficult to get an accurate sense of what work needs to be done on future releases. With JupyterLab, we are using Backlog for all future issues, and beta, 1.0 only for issues that we have more-or-less committed to implementing for those releases. This way our beta and 1.0 milestones are reasonably accurate.

The other traditional role of the wishlist milestone (unstated, but in practice) is that it was where we put community feature requests we were sure we were going to implement, but didn't want to outright close. In the end "wishlist" ended up being the place where issues went to die. The best example of that is the ipython/ipython repo with has >600 issues with the wishlist label:

https://github.com/ipython/ipython/issues?q=is%3Aopen+is%3Aissue+milestone%3Awishlist

By using a single Backlog label for all future work in JupyterLab (our own priorities and community requests) we feel we are better able to treat community requests in the same way we treat our own priorities.

I will try to clarify this in the PR...